packages/backend-cli/src/directLine.ts

```diff
@@ -0,0 +1,772 @@
+          this.connectionStatus$.next(ConnectionStatus.Online);
+          return Observable.of(connectionStatus);
+        } else {
+          return this.startConversation().do(conversation => {
+            this.conversationId = conversation.conversationId;
+            this.token = this.secret || conversation.token;
+            this.streamUrl = conversation.streamUrl;
+            this.referenceGrammarId = conversation.referenceGrammarId;
+            if (!this.secret)
+              this.refreshTokenLoop();
+
+            this.connectionStatus$.next(ConnectionStatus.Online);
+          }, error => {
+            this.connectionStatus$.next(ConnectionStatus.FailedToConnect);
+          })
+          .map(_ => connectionStatus);
+        }
+      }
+      else {
+        return Observable.of(connectionStatus);
+      }
+    })
+    .filter(connectionStatus => connectionStatus != ConnectionStatus.Uninitialized && connectionStatus != ConnectionStatus.Connecting)
+    .flatMap(connectionStatus => {
+      switch (connectionStatus) {
+        case ConnectionStatus.Ended:
+          return Observable.throw(errorConversationEnded);
+
+        case ConnectionStatus.FailedToConnect:
+          return Observable.throw(errorFailedToConnect);
+
+        case ConnectionStatus.ExpiredToken:
+          return Observable.throw(errorExpiredToken);
+
+        default:
+          return Observable.of(null);
+      }
+    })
+
+    return once ? obs.take(1) : obs;
+  }
+
+  private expiredToken() {
+    const connectionStatus = this.connectionStatus$.getValue();
+    if (connectionStatus != ConnectionStatus.Ended && connectionStatus != ConnectionStatus.FailedToConnect)
+      this.connectionStatus$.next(ConnectionStatus.ExpiredToken);
+  }
+
+  private startConversation() {
+    //if conversationid is set here, it means we need to call the reconnect api, else it is a new conversation
+    const url = this.conversationId
+      ? `${this.domain}/conversations/${this.conversationId}?watermark=${this.watermark}`
+      : `${this.domain}/conversations`;
+    const method = this.conversationId ? "GET" : "POST";
+
+    return this.fetchAsObservable({
+      method,
+      timeout,
+      url,
+      headers: {
+        "Accept": "application/json",
+        "Authorization": `Bearer ${this.token}`
+      }
+    })
+//      .do(ajaxResponse => konsole.log("conversation ajaxResponse", ajaxResponse.response))
+    .map(ajaxResponse => ajaxResponse.response as Conversation)
+    .retryWhen(error$ =>
+      // for now we deem 4xx and 5xx errors as unrecoverable
+      // for everything else (timeouts), retry for a while
+      error$.mergeMap(error => !error.status || (error.status >= 400 && error.status < 600)
+        ? Observable.throw(error)
+        : Observable.of(error)
+      )
+      .delay(timeout)
+      .take(retries)
+    )
+  }
+
+  private refreshTokenLoop() {
+    this.tokenRefreshSubscription = Observable.interval(intervalRefreshToken)
```

Comment:
if the subscription is already created do you want to skip this - do you want to unsubscribe/release after it is refreshed?

Reply:
It looks like it is called in checkconnection which is called often.

Reply:
and called from refreshtoken()